samples/chat-demo/app.go

```diff
@@ -49,89 +82,186 @@ func (chat *ChatApp) ApplyEvents(eventsIn *events.EventList) (*events.EventList,
+// The snapshot is a binary representation of the application state that can be passed to applyRestoreState().
+func (chat *ChatApp) applySnapshotRequest(snapshotRequest *eventpb.StateSnapshotRequest) (*events.EventList, error) {
+	return events.ListOf(events.StateSnapshotResponse(
+		t.ModuleID(snapshotRequest.Module),
+		events.StateSnapshot(chat.serializeMessages(), events.EpochConfig(chat.currentEpoch, chat.memberships)),
+	)), nil
+}
+
+func (chat *ChatApp) serializeMessages() []byte {
+	data := make([]byte, 0)
+	for _, msg := range chat.messages {
+		data = append(data, []byte(msg)...)
+		data = append(data, 0)
+	}
+
+	return data
 }
 
 // RestoreState restores the application's state to the one represented by the passed argument.
 // The argument is a binary representation of the application state returned from Snapshot().
 // After the chat history is restored, RestoreState prints the whole chat history to stdout.
-func (chat *ChatApp) RestoreState(snapshot []byte) error {
+func (chat *ChatApp) applyRestoreState(snapshot *commonpb.StateSnapshot) (*events.EventList, error) {
 
-	// Unmarshal the protobuf message from its binary form.
-	state := &AppState{}
-	if err := proto.Unmarshal(snapshot, state); err != nil {
-		return err
-	}
+	// Restore chat messages
+	chat.restoreChat(snapshot.AppData)
 
-	// Restore internal state
-	chat.messages = state.Messages
+	// Restore configuration
+	if err := chat.restoreConfiguration(snapshot.Configuration); err != nil {
+		return nil, err
+	}
 
 	// Print new state
-	fmt.Printf("\n CHAT STATE RESTORED. SHOWING ALL CHAT HISTORY FROM THE BEGINNING.\n")
+	fmt.Printf("\nCHAT STATE RESTORED. SHOWING ALL CHAT HISTORY FROM THE BEGINNING.\n\n")
 	for _, message := range chat.messages {
 		fmt.Println(message)
 	}
 
-	return nil
+	return events.EmptyList(), nil
 }
 
-// The ImplementsModule method only serves the purpose of indicating that this is a Module and must not be called.
-func (chat *ChatApp) ImplementsModule() {}
+func (chat *ChatApp) restoreChat(data []byte) {
+	chat.messages = make([]string, 0)
+	if len(data) > 0 {
+		for _, msg := range bytes.Split(data[:len(data)-1], []byte{0}) { // len(data)-1 to strip off the last zero byte
+			chat.messages = append(chat.messages, string(msg))
+		}
+	}
+}
+
+func (chat *ChatApp) restoreConfiguration(config *commonpb.EpochConfig) error {
+	chat.currentEpoch = t.EpochNr(config.EpochNr)
+	chat.memberships = make([]map[t.NodeID]t.NodeAddress, len(config.Memberships))
+	for e, mem := range config.Memberships {
+		chat.memberships[e] = make(map[t.NodeID]t.NodeAddress)
+		for nID, nAddr := range mem.Membership {
+			var err error
+			chat.memberships[e][t.NodeID(nID)], err = multiaddr.NewMultiaddr(nAddr)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	fmt.Printf("Restored app memberships: %d (epoch: %d)\n", len(chat.memberships), chat.currentEpoch)
```

Comment:
debug output, remove or add TODO to remove it

Reply:
Fixed.